File: lua/tests/test-nibs-encoder.lua
local Utils = require 'test-utils'
local Nibs = require "nibs"
local Tibs = require 'tibs'
local NibLib = require 'nib-lib'

local colorize = require('pretty-print').colorize
local readFileSync = require('fs').readFileSync

local filename = module.dir .. "/../../fixtures/encoder-fixtures.tibs"
local text = assert(readFileSync(filename))
local allTests = assert(Tibs.decode(text))

for _ = 1, 10 do -- Multiple runs to exercise GC more
    collectgarbage("collect")
    for section, tests in pairs(allTests) do
        print("\n" .. colorize("highlight", section) .. "\n")
        for i = 1, #tests, 2 do
            local input = tests[i]
            local expected = NibLib.bufToStr(tests[i + 1])
            local actual = Nibs.encode(input)

            print(string.format("% 40s → %s",
                Tibs.encode(input),
                colorize(expected == actual and "success" or "failure", NibLib.strToHexStr(actual))))
            if expected ~= actual then
                collectgarbage("collect")
                print(colorize("failure", string.format("% 40s | %s",
                    "Error, not as expected",
                    colorize("success", NibLib.strToHexStr(expected)))))
                return nil, "Encode Mismatch"
            end
            collectgarbage("collect")
        end
        collectgarbage("collect")
    end
    collectgarbage("collect")
end

File: js/tests/test-nibs-encoder.js
import { readFileSync } from 'fs'
import { join as pathJoin } from 'path'
import { fileURLToPath } from 'url';

import * as Tibs from '../tibs.js'
import * as Nibs from '../nibs.js'
import { inspect } from 'util';

const filename = pathJoin(fileURLToPath(import.meta.url), '../../../fixtures/encoder-fixtures.tibs')
/** @type {Map} */
const tests = Tibs.decode(readFileSync(filename, 'utf8'), filename)

function highlight(str) {
    return `\x1b[38;5;45;48;5;236m${str}\x1b[0m`
}

/**
 * 
 * @param {Uint8Array} a 
 * @param {Uint8Array} b 
 * @returns {boolean}
 */
function same(a, b) {
    if (a.length !== b.length) return false
    for (let i = 0, l = a.length; i < l; i++) {
        if (a[i] !== b[i]) return false
    }
    return true
}

for (const [name, list] of tests.entries()) {
    console.log(`\n${highlight(name)}`)
    for (let i = 0; i < list.length; i += 2) {
        const input = list[i];
        const expected = list[i + 1]
        const actual = Nibs.encode(input)
        console.log("\nInput    " + inspect(input, { showHidden: true, depth: 2, colors: true, breakLength: 100 }))
        console.log("Expected " + Buffer.from(expected).toString('hex'))
        console.log("Actual   " + Buffer.from(actual).toString('hex'))
        if (!same(expected, actual)) {
            throw new Error("Mismatch")
        }
    }
}
